### DISCUSSION TITLE
Solve coding problems outside English

### DISCUSSION THREAD

**Discussion by yibo7** [CoderLine/alphaTab#638] [Q&A]
In the old version, I solved the problem of garbled code like this：

C#:
```
public static string ToString(byte[] data)
        {
            //return Encoding.UTF8.GetString(data);
            return Encoding.Default.GetString(data);
        }
```
js:
```
AlphaTab.Platform.Std.ToString = function (data){
    //... old code
    var s = new TextDecoder('gbk').decode(data);      
    return s;
};
```
How to solve it in the new version

**Comment by Danielku15** [comment]
You can control the encoding via [`importer.encoding`](https://alphatab.net/docs/reference/settings/importer/encoding#description) setting.

**Reply by yibo7** [reply]
There is still a small problem in the C# project:

``` csharp

var settings = new Settings();
            settings.Core.Engine = "skia";
            settings.Importer.Encoding = "gb2312";
            // load score
            var score = ScoreLoader.LoadScoreFromBytes(File.ReadAllBytes(sPath), settings);

            // render score with svg engine and desired rendering width
          
            var renderer = new ScoreRenderer(settings)
            {
                Width = 970
            };
```


In the above code, "score" can get the result correctly, but the "renderer" still has Chinese garbled characters

**Reply by Danielku15** [reply]
What file formats are you using? Do you have an example file showing the issue?

**Reply by yibo7** [reply]
![image](https://user-images.githubusercontent.com/58937236/130220758-37052c88-1a83-4d13-b82f-9a628f257805.png)
The files I use are. GP3,. Gp4. They can be encoded correctly in HTML, but they can't be encoded correctly in c# Chinese

**Reply by yibo7** [reply]
**This is one of my test files**


[test-file.zip](https://github.com/CoderLine/alphaTab/files/7020962/test-file.zip)

**Reply by Danielku15** [reply]
I had a closer look and the issue seems to come from the fact that the default fonts used in alphaTab do not have chinese characters in it. 
You will need to reconfigure alphaTab to use a font which has chinese characters in it using the [resources](https://alphatab.net/docs/reference/settings/display/resources) option. The fonts need to be installed in the system because there is no public way to register custom font files (yet). You could by pass this with reflection by calling this internal method: 
https://github.com/CoderLine/alphaTab/blob/70eb7c3d99ad9d02c4e3b5fc93a8523a8ed33e78/src.csharp/AlphaTab/Platform/CSharp/SkiaCanvas.cs#L68 

I gave it a test to change the title font to "SimSum" (default Windows font for Chinese characters) and it displays the characters correctly:

```cs
settings.Display.Resources.TitleFont = new Font("SimSun",
    settings.Display.Resources.TitleFont.Size,
    settings.Display.Resources.TitleFont.Style,
    settings.Display.Resources.TitleFont.Weight);
```

![image](https://user-images.githubusercontent.com/674916/130350401-c6d338c7-1ff5-4292-8b08-0bc8ccf40ef4.png)

It seems browsers have some advanced fallback mechanism to change the font on character level.

**Reply by yibo7** [reply]
Many thanks for your kind and warm help.
